libraries/chain/asset_evaluator.cpp

```diff
@@ -352,8 +358,9 @@ void_result asset_update_issuer_evaluator::do_evaluate(const asset_update_issuer
 
    if( d.head_block_time() < HARDFORK_CORE_199_TIME )
    { // TODO: remove after HARDFORK_CORE_199_TIME has passed
-      graphene::chain::impl::hf_199_visitor hf_199;
-      hf_199( o );
+      graphene::chain::impl::hf_visitor hf_visitor;
+      hf_visitor.block_time = d.head_block_time();
+      hf_visitor( o );
```

Comment:
Here, I think it's better to not visit the generic visitor but use `FC_ASSERT` directly.

Actually, to make the visitor more generic, I think it's better to organize the code like this: add a generic member function in the base class e.g. `generic_evaluator`, say, name it `simple_hardfork_check()`, call it in `evaluate()` of base class, override it in derived classes when needed, then cleanup when hard fork time has passed. Perhaps the scope is a bit too wide?

Reply:
what if we rename `operation_hardfork_visitor` to `proposal_hardfork_visitor`, create a new `asset_hardfork_visitor` in `asset_evaluator.hpp` and do the asset hardfork logics in there as we did for proposal?

the only thing i can see is that in the case of asset there is also hardfork code(`HARDFORK_CORE_429_TIME`) in `do_apply` that we will need to consider or leave out.

Reply:
>what if we rename `operation_hardfork_visitor` to `proposal_hardfork_visitor`, create a new `asset_hardfork_visitor` in `asset_evaluator.hpp` and do the asset hardfork logics in there as we did for proposal?

We can do it with one line of code, why need a visitor? Rename if fine to me.

>the only thing i can see is that in the case of asset there is also hardfork code(`HARDFORK_CORE_429_TIME`) in `do_apply` that we will need to consider or leave out.

This is what I said "scope too wide". It will make things unnecessarily complicated.

Reply:
then we just leave it as it is(at least for now) for asset hardfork asserts ?

Reply:
Yes, keep it for now.